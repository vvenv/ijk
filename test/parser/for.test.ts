import { describe, expect, test } from 'vitest';
import { parse } from './__helper';

describe('validation', () => {
  test('start and end tag mismatched', () => {
    expect(() => parse('{{ if x }}{{ endfor }}')).toMatchSnapshot();
    expect(() =>
      parse('{{ if x }}{{ endfor }}', { debug: true }),
    ).toThrowErrorMatchingSnapshot();
    expect(() => {
      try {
        parse('{{ if x }}{{ endfor }}', { debug: true });
      } catch (error: any) {
        throw new Error(error.details);
      }
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('for - array', () => {
  test('basic', () => {
    expect(
      parse(
        `{{ names }}{{ for name of names }}{{ name }} of {{ names }}{{ endfor }}{{ names }}`,
      ),
    ).toMatchSnapshot();
  });

  test('destructuring', () => {
    expect(
      parse(`{{ for x,y,z of a }}{{ x }},{{ y }},{{ z }}{{ endfor }}`),
    ).toMatchSnapshot();
  });

  test('duo', () => {
    expect(
      parse(
        `{{ for name of names }}{{ name }} of {{ names }}{{ endfor }}{{ for name of names }}{{ name }} of {{ names }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('nesting', () => {
    expect(
      parse(
        `{{ for as of ass }}{{ for a of as }}{{ a }} of {{ as }} of {{ ass }}{{ endfor }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('loop.index', () => {
    expect(
      parse(
        `{{ for name of names }}{{loop.index+1}} {{ name }} of {{ names }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('else', () => {
    expect(
      parse(`{{ for name of names }}{{ name }}{{ else }}empty{{ endfor }}`),
    ).toMatchSnapshot();
  });

  test('filter', () => {
    expect(
      parse(
        `{{ for name of names | split }}{{ name }} of {{ names }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('filter w/ params', () => {
    expect(
      parse(
        `{{ for char of name | split "" }}{{ char }} of {{ name }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });
});

describe('for - object', () => {
  test('basic', () => {
    expect(
      parse(
        `{{ names }}{{ for name in names }}{{ name }} in {{ names }}{{ endfor }}{{ names }}`,
      ),
    ).toMatchSnapshot();
  });

  test('duo', () => {
    expect(
      parse(
        `{{ for name in names }}{{ name }} in {{ names }}{{ endfor }}{{ for name in names }}{{ name }} in {{ names }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('nesting', () => {
    expect(
      parse(
        `{{ for as in ass }}{{ for a in as }}{{ a }} in {{ as }} in {{ ass }}{{ endfor }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('loop.index', () => {
    expect(
      parse(
        `{{ for name in names }}{{loop.index+1}} {{ name }} in {{ names }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('else', () => {
    expect(
      parse(`{{ for name in names }}{{ name }}{{ else }}empty{{ endfor }}`),
    ).toMatchSnapshot();
  });

  test('filter', () => {
    expect(
      parse(
        `{{ for name in names | split }}{{ name }} in {{ names }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });

  test('filter w/ params', () => {
    expect(
      parse(
        `{{ for char in name | split "" }}{{ char }} in {{ name }}{{ endfor }}`,
      ),
    ).toMatchSnapshot();
  });
});
