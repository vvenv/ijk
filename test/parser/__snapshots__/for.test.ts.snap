// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`for - array > basic 1`] = `""use strict";let s="";s+=e(c.names);const o_1_1=c.names;const a_1_1=Array.isArray(o_1_1);const k_1_1=Object.keys(o_1_1);const l_1_1=k_1_1.length;for(let i_1_1=0;i_1_1<l_1_1;i_1_1++){const c_i_1_1={...c,name:o_1_1[k_1_1[i_1_1]],loop:{index:i_1_1,first:i_1_1===0,last:i_1_1===l_1_1,length:l_1_1}};s+=e(c_i_1_1.name);s+=" in ";s+=e(c_i_1_1.names);}s+=e(c.names);return s;"`;

exports[`for - array > constructing 1`] = `""use strict";let s="";const o_1_0=[a, b, c];const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,x:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.x);}return s;"`;

exports[`for - array > destructuring 1`] = `""use strict";let s="";const o_1_0=c.a;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,x:a_1_0?o_1_0[k_1_0[i_1_0]][0]:0===0?k_1_0[i_1_0]:o_1_0[k_1_0[i_1_0]],y:a_1_0?o_1_0[k_1_0[i_1_0]][1]:1===0?k_1_0[i_1_0]:o_1_0[k_1_0[i_1_0]],z:a_1_0?o_1_0[k_1_0[i_1_0]][2]:2===0?k_1_0[i_1_0]:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.x);s+=",";s+=e(c_i_1_0.y);s+=",";s+=e(c_i_1_0.z);}return s;"`;

exports[`for - array > duo 1`] = `""use strict";let s="";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.name);s+=" in ";s+=e(c_i_1_0.names);}const o_1_1=c.names;const a_1_1=Array.isArray(o_1_1);const k_1_1=Object.keys(o_1_1);const l_1_1=k_1_1.length;for(let i_1_1=0;i_1_1<l_1_1;i_1_1++){const c_i_1_1={...c,name:o_1_1[k_1_1[i_1_1]],loop:{index:i_1_1,first:i_1_1===0,last:i_1_1===l_1_1,length:l_1_1}};s+=e(c_i_1_1.name);s+=" in ";s+=e(c_i_1_1.names);}return s;"`;

exports[`for - array > else 1`] = `""use strict";let s="";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;if(l_1_0){for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.name);}}else{s+="empty";}return s;"`;

exports[`for - array > filter 1`] = `""use strict";let s="";const o_1_0=f.split(c.names);const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.name);s+=" in ";s+=e(c_i_1_0.names);}return s;"`;

exports[`for - array > filter w/ params 1`] = `""use strict";let s="";const o_1_0=f.split(c.name,"");const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,char:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.char);s+=" in ";s+=e(c_i_1_0.name);}return s;"`;

exports[`for - array > loop.index 1`] = `""use strict";let s="";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.loop.index+1);s+=" ";s+=e(c_i_1_0.name);s+=" in ";s+=e(c_i_1_0.names);}return s;"`;

exports[`for - array > nesting 1`] = `""use strict";let s="";const o_1_0=c.ass;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,as:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};const o_2_0=c_i_1_0.as;const a_2_0=Array.isArray(o_2_0);const k_2_0=Object.keys(o_2_0);const l_2_0=k_2_0.length;for(let i_2_0=0;i_2_0<l_2_0;i_2_0++){const c_i_1_0_i_2_0={...c_i_1_0,a:o_2_0[k_2_0[i_2_0]],loop:{index:i_2_0,first:i_2_0===0,last:i_2_0===l_2_0,length:l_2_0}};s+=e(c_i_1_0_i_2_0.a);s+=" in ";s+=e(c_i_1_0_i_2_0.as);s+=" in ";s+=e(c_i_1_0_i_2_0.ass);}}return s;"`;

exports[`for - object > basic 1`] = `""use strict";let s="";s+=e(c.names);const o_1_1=c.names;const a_1_1=Array.isArray(o_1_1);const k_1_1=Object.keys(o_1_1);const l_1_1=k_1_1.length;for(let i_1_1=0;i_1_1<l_1_1;i_1_1++){const c_i_1_1={...c,name:o_1_1[k_1_1[i_1_1]],loop:{index:i_1_1,first:i_1_1===0,last:i_1_1===l_1_1,length:l_1_1}};s+=e(c_i_1_1.name);s+=" in ";s+=e(c_i_1_1.names);}s+=e(c.names);return s;"`;

exports[`for - object > destructuring 1`] = `""use strict";let s="";const o_1_0=c.a;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,k:a_1_0?o_1_0[k_1_0[i_1_0]][0]:0===0?k_1_0[i_1_0]:o_1_0[k_1_0[i_1_0]],v:a_1_0?o_1_0[k_1_0[i_1_0]][1]:1===0?k_1_0[i_1_0]:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.k);s+=":";s+=e(c_i_1_0.v);}return s;"`;

exports[`for - object > duo 1`] = `""use strict";let s="";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.name);s+=" in ";s+=e(c_i_1_0.names);}const o_1_1=c.names;const a_1_1=Array.isArray(o_1_1);const k_1_1=Object.keys(o_1_1);const l_1_1=k_1_1.length;for(let i_1_1=0;i_1_1<l_1_1;i_1_1++){const c_i_1_1={...c,name:o_1_1[k_1_1[i_1_1]],loop:{index:i_1_1,first:i_1_1===0,last:i_1_1===l_1_1,length:l_1_1}};s+=e(c_i_1_1.name);s+=" in ";s+=e(c_i_1_1.names);}return s;"`;

exports[`for - object > else 1`] = `""use strict";let s="";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;if(l_1_0){for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.name);}}else{s+="empty";}return s;"`;

exports[`for - object > filter 1`] = `""use strict";let s="";const o_1_0=f.split(c.names);const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.name);s+=" in ";s+=e(c_i_1_0.names);}return s;"`;

exports[`for - object > filter w/ params 1`] = `""use strict";let s="";const o_1_0=f.split(c.name,"");const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,char:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.char);s+=" in ";s+=e(c_i_1_0.name);}return s;"`;

exports[`for - object > loop.index 1`] = `""use strict";let s="";const o_1_0=c.names;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,name:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};s+=e(c_i_1_0.loop.index+1);s+=" ";s+=e(c_i_1_0.name);s+=" in ";s+=e(c_i_1_0.names);}return s;"`;

exports[`for - object > nesting 1`] = `""use strict";let s="";const o_1_0=c.ass;const a_1_0=Array.isArray(o_1_0);const k_1_0=Object.keys(o_1_0);const l_1_0=k_1_0.length;for(let i_1_0=0;i_1_0<l_1_0;i_1_0++){const c_i_1_0={...c,as:o_1_0[k_1_0[i_1_0]],loop:{index:i_1_0,first:i_1_0===0,last:i_1_0===l_1_0,length:l_1_0}};const o_2_0=c_i_1_0.as;const a_2_0=Array.isArray(o_2_0);const k_2_0=Object.keys(o_2_0);const l_2_0=k_2_0.length;for(let i_2_0=0;i_2_0<l_2_0;i_2_0++){const c_i_1_0_i_2_0={...c_i_1_0,a:o_2_0[k_2_0[i_2_0]],loop:{index:i_2_0,first:i_2_0===0,last:i_2_0===l_2_0,length:l_2_0}};s+=e(c_i_1_0_i_2_0.a);s+=" in ";s+=e(c_i_1_0_i_2_0.as);s+=" in ";s+=e(c_i_1_0_i_2_0.ass);}}return s;"`;

exports[`validation > start and end tag mismatched 1`] = `[Function]`;

exports[`validation > start and end tag mismatched 2`] = `[ASTError: "endfor" must follow "for", not "if".]`;

exports[`validation > start and end tag mismatched 3`] = `
[Error:  i>j>k  "endfor" must follow "for", not "if".

1: {{ if x }}{{ endfor }}
   ^^^^^^^^^^
             ^^^^^^^^^^^^
]
`;
